### DISCUSSION TITLE
PIMCore: DataHub REST : Retrieving records when record size is more than 10000

### DISCUSSION THREAD

**Discussion by ShaileshT-dot** [pimcore/pimcore#11462] [Q&A / Support] [answered]
1.	We are having ElasticSearch (7.15.2) installed to support DataHub REST APIs in our PIMCore Enterprise Edition. The configuration of the ES Server is completed and Products do get indexed without any issues.
2.	In one of the index, we have around 39K records.
3.	We have to query this ES index and get all product details, eventually. It does not mean we need all product details in one single request. It is ok for us to get product details in multiple batches, batch size of 1000.
4.	For achieving this PIMcore DataHub REST provides two parameters: size and page_cursor. Size specifies the size of the batch and page_cursor the batch number(or start of the index number). However, this is not working as expected.
5.	ES has OOB index.max_result_window which is set to 10000, default value. We don’t intend to change this value.
6.	When using size=1000 and page_cursor is 10001, when we try to fetch the record from 10001 onwards, we get an error message stating: “Result window exceeds 10000 records”. 
7.	One of the solution for this is to use search_after as a workaround. Using "search_after" parameter we have been able to get the required results if we query on the ElasticSearch Index 
8. However, if we are using "search_after" parameter while querying on DataHub REST API Endpoint, we are not able to get any results. Probably, this could be because PIMCore is not recognizing the "search_after" parameter. Since PIMCore is not recognizing "search_after" parameter , it is not passing this parameter internally to ElasticSearch, hence we are not getting the desired results.
9. Question: Can PIMcore provide a resolution on how to get the required results, when the index size is more than 10000 records.
10. Question: Can PIMCore provide a resolution on usage of "search_after" parameter?
11. One of the solution could be to query the ElasticSearch Index directly and not go via the PIMCore REST End point, but then it will not have the authentication mechanisms which PIMCore provides. However, I'm seeking better resolution from PIMCore. Thanks.

**Comment by fashxp** [comment, marked as answer]
when loading the next page of a result, please use for the `page_cursor` parameter the value provided by the previous response ... 
either by the link header, or the data attribute `page_cursor` in the response. 

![image](https://user-images.githubusercontent.com/8792145/154927164-2d75fe36-d668-45fd-94e9-d53bd62d3313.png)


the `page_cursor` has two operation modes 
- for pages below the configured max results window, it uses a numeric value and ES `from` query
- for pages after the max results window, it uses the `search_after` option of ES - as also [here suggested](https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html).

**Reply by ShaileshT-dot** [reply]
Hi Christian,
Thanks for your detailed response. 
As mentioned in my original post, we have tried with search_after parameter, but page_cursor works well to some extent. please see point 4 in my original post.
As per point 8 and 10 in my post, search_after does not work when we query for the results via the PIMCore REST API. 
search_after works well, as said in the link you shared, only if we are querying directly on the ES index. It does not work when we query on the PIMCore REST Endpoint.

**Reply by fashxp** [reply]
just tested this with the simple rest api from (screenshots are from the swagger ui of it) ... and it works as described.

**Reply by ShaileshT-dot** [reply]
verified it. thanks for your support.
